File: testbed_ruby/test/associations/optional_one_to_many_test.rb
require 'test/unit'
require 'src/student_c'
require 'src/mentor_c'
require 'src/program_c'

module CruiseTest
class OptionalOneToManyTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentC.new(99)
    assert_equal(nil,s.get_mentor)
  end
  
  def test_add_student
    m = MentorC.new("m1")
    s = StudentC.new(99)
    m.add_student(s)
    
    assert_equal(m,s.get_mentor)
    assert_equal(s,m.get_student(0))
  end
  
  def test_replaceMentor
    m = MentorC.new("m1")
    m2 = MentorC.new("m2")

    s = StudentC.new(99)
    m.add_student(s)
    assert_equal(1,m.number_of_students)
 
    s.set_mentor(m2)
    
    assert_equal(m2,s.get_mentor)
    assert_equal(s,m2.get_student(0))
    assert_equal(0,m.number_of_students)
  end 
  
  def test_addToNewMentor
    m = MentorC.new("m1")
    m2 = MentorC.new("m2")
    s = StudentC.new(99)
    
    m.add_student(s)
    m2.add_student(s)
    
    assert_equal(m2,s.get_mentor)
    assert_equal(s,m2.get_student(0))
    assert_equal(0,m.number_of_students)
  end  
  
  def test_removeFromExistingMentor
    m = MentorC.new("m1")
    m2 = MentorC.new("m2")

    s1 = StudentC.new(99)
    s2 = StudentC.new(98)
    s3 = StudentC.new(97)
    
    m.add_student(s1)
    m.add_student(s2)
    m.add_student(s3)

    assert_equal(3,m.number_of_students)
    assert_equal(0,m2.number_of_students)
    
    m2.add_student(s1)

    assert_equal(2,m.number_of_students)
    assert_equal(s2,m.get_student(0))
    assert_equal(s3,m.get_student(1))
    
    assert_equal(1,m2.number_of_students)
    assert_equal(s1,m2.get_student(0))
    
    assert_equal(m2,s1.get_mentor)
    assert_equal(m,s2.get_mentor)
    assert_equal(m,s3.get_mentor)
  end   
  
  def test_remove_student
    m = MentorC.new("m1")
    s = StudentC.new(99)
    
    m.add_student(s)
    m.remove_student(s)
    
    assert_equal(nil,s.get_mentor)
    assert_equal(0,m.number_of_students)
  end  
  
end
end

File: testbed_python/test/associations/optional_one_to_many_test.py
import unittest

# Test classes are imported dynamically
# Must be generated into ImportModules' namespace then imported
from ImportModules import importModules

importModules(["MentorC", "StudentC"], ["associations"])
from ImportModules import *


class OptionalOneToManyTest(unittest.TestCase):
    def test_createStudentWithoutMentor(self):
        s = StudentC.StudentC(99)
        self.assertEqual(None, s.getMentor())

    def test_addStudent(self):
        m = MentorC.MentorC("m1")
        s = StudentC.StudentC(99)
        m.addStudent(s)

        self.assertEqual(1, m.numberOfStudents())
        self.assertEqual(m, s.getMentor())
        self.assertEqual(s, m.getStudent(0))

    def test_replaceMentor(self):
        m = MentorC.MentorC("m1")
        m2 = MentorC.MentorC("m2")

        s = StudentC.StudentC(99)
        m.addStudent(s)
        self.assertEqual(1, m.numberOfStudents())
        s.setMentor(m2)

        self.assertEqual(m2, s.getMentor())
        self.assertEqual(s, m2.getStudent(0))
        self.assertEqual(0, m.numberOfStudents())

    def test_addToNewMentor(self):
        m = MentorC.MentorC("m1")
        m2 = MentorC.MentorC("m2")
        s = StudentC.StudentC(99)

        m.addStudent(s)
        m2.addStudent(s)

        self.assertEqual(m2, s.getMentor())
        self.assertEqual(s, m2.getStudent(0))
        self.assertEqual(0, m.numberOfStudents())

    def test_removeFromExistingMentor(self):
        m = MentorC.MentorC("m1")
        m2 = MentorC.MentorC("m2")

        s1 = StudentC.StudentC(99)
        s2 = StudentC.StudentC(98)
        s3 = StudentC.StudentC(97)

        m.addStudent(s1)
        m.addStudent(s2)
        m.addStudent(s3)

        self.assertEqual(3, m.numberOfStudents())
        self.assertEqual(0, m2.numberOfStudents())

        m2.addStudent(s1)

        self.assertEqual(2, m.numberOfStudents())
        self.assertEqual(s2, m.getStudent(0))
        self.assertEqual(s3, m.getStudent(1))

        self.assertEqual(1, m2.numberOfStudents())
        self.assertEqual(s1, m2.getStudent(0))

        self.assertEqual(m2, s1.getMentor())
        self.assertEqual(m, s2.getMentor())
        self.assertEqual(m, s3.getMentor())

    def test_removeStudent(self):
        m = MentorC.MentorC("m1")
        s = StudentC.StudentC(99)

        m.addStudent(s)
        m.removeStudent(s)

        self.assertEqual(None, s.getMentor())
        self.assertEqual(0, m.numberOfStudents())

    def test_deleteMentor(self):
        m = MentorC.MentorC("m1")
        s1 = StudentC.StudentC(1)
        s2 = StudentC.StudentC(2)
        s3 = StudentC.StudentC(3)
        s1.setMentor(m)
        s2.setMentor(m)
        s3.setMentor(m)

        del m

        self.assertFalse(s1.getMentor() in locals())
        self.assertFalse(s2.getMentor() in locals())
        self.assertFalse(s3.getMentor() in locals())
